detox/src/ipc/ipc.test.js

```diff
@@ -17,97 +20,164 @@ describe('IPC', () => {
+
+      it('when called directly, should register a worker', async () => {
+        await ipcServer.onRegisterWorker({ workerId: 'worker-1' });
+        expect(ipcServer.sessionState.workersCount).toEqual(1);
 
-      const disposePromise = ipcServer.dispose();
-      disposePromise.then(() => {
-        fail('dispose() should not resolve before all clients are disconnected');
+        await ipcServer.onRegisterWorker({ workerId: 'worker-2' });
+        expect(ipcServer.sessionState.workersCount).toEqual(2);
       });
 
-      await new Promise(resolve => setTimeout(resolve, 3000));
+      it('should not count the same worker twice', async () => {
+        await ipcServer.onRegisterWorker({ workerId: 'worker-1' });
+        expect(ipcServer.sessionState.workersCount).toEqual(1);
+
+        await ipcServer.onRegisterWorker({ workerId: 'worker-1' });
+        expect(ipcServer.sessionState.workersCount).toEqual(1);
+      });
     });
 
-    it.skip('should resolve dispose() if no clients are connected', async () => {
-      setTimeout(() => {
-        throw new Error('dispose() should have resolved by now');
-      }, 3000);
+    describe('onReportTestResults', () => {
+      beforeEach(() => ipcServer.init());
+
+      const success = (testFilePath) => ({ testFilePath, success: true });
+      const failure = (testFilePath) => ({ testFilePath, success: false, testExecError: { message: 'error' } });
 
-      await ipcServer.dispose();
+      it('when called directly, should add test results', async () => {
+        const testResults = [success('test-1'), failure('test-2')];
+        await ipcServer.onReportTestResults({ testResults });
+        expect(ipcServer.sessionState.testResults).toEqual(testResults);
+      });
+
+      it('when called more than one time, should merge test results', async () => {
+        const testResults = [failure('test-1'), success('test-2')];
+        await ipcServer.onReportTestResults({ testResults });
+        await ipcServer.onReportTestResults({ testResults: [success('test-1')] });
+
+        expect(ipcServer.sessionState.testResults).toEqual([
+          success('test-1'),
+          success('test-2'),
+        ]);
+      });
     });
 
-    it.skip('should resolve dispose() if all clients have disconnected', async () => {
-      await ipcClient1.init();
-      await ipcClient2.init();
+    describe('dispose()', () => {
+      it('should resolve if there are no connected clients', async () => {
+        await ipcServer.init();
+        await expect(ipcServer.dispose()).resolves.toBeUndefined();
+      });
 
-      await ipcClient1.dispose();
-      await ipcClient2.dispose();
+      it('should wait for connected clients to disconnect', async () => {
+        await ipcServer.init();
+        await ipcClient1.init();
 
-      setTimeout(() => {
-        throw new Error('dispose() should have resolved by now');
-      }, 3000);
+        const disposePromise = ipcServer.dispose();
```

Comment:
I decided to make this test complete.
First, we wait to make sure that there is no magic, and that the IPC server doesn't actually close.
Then we close the client and make sure that the server closes.